### PR TITLE
fix: user mention beside a special character

### PR DESCRIPTION
**Pull request by sshanzel**
I have written numerous test cases to ensure all the edge cases will be covered. The behavior of when to allow the mention of what special character it was beside, was based on Twitter. On Twitter, they allow any special character so long as it is not the at (@) sign.

First, we replace the special characters with spaces to figure out which ones are valid mentions of users.

Then while we check each word for any mention, we replace again the special characters but this time we store them in an array so we can easily reconstruct the word for which one was replaced.

If the mention was found we will generate the anchor tag with it.

### CODE REVIEW

**Comment by idoshamun** [comment]
It's somewhat hard to understand the logic you describe here in your code.
Another suggestion regarding your algorithm that might ease the process is before you split the original text by spaces, you can replace all special characters with spaces and then do the split. At the end of the process, you can change the relevant spaces to the special characters. This way you will not have to deal with them in the middle of the process because you can assume they are spaces

**Comment by sshanzel** [comment]
I think that can really work, the only problem we have is when we are going to map the replaced special characters to the right values after processing because the content was reconstructed and doesn't have the same length anymore. We would again have to split it by special characters to make it an array so we can track the position of the special characters it found. WDYT?

**Comment by idoshamun** [comment]
I'll leave it to you to decide what's the best approach, if you want to talk about I'm available today. Either way, we need to make the implementation more readable and maintainable

**Comment by sshanzel** [comment]
@idoshamun I really agreed with your comment about making it much simpler. I applied what you suggested though I had to make the replacement while processing the word to make reconstructing the replaced words much easier (mapping). WDYT with the current implementation?

File: src/common/markdown.ts [hunk 1]
```diff
@@ -43,16 +43,49 @@ const defaultTextRender = markdown.renderer.rules.text;
...
+type ReplacedCharacters = string[];
+
+const getReplacedCharacters = (word: string): [string, ReplacedCharacters] => {
+  const specialCharacters = [];
+  word.split('').forEach((char) => {
+    const matched = char.match(mentionSpecialCharacters);
+    if (matched?.[0]?.length) {
+      specialCharacters.push(char);
+    }
+  });
```
Comment (idoshamun) [inline src/common/markdown.ts:57]: You can use something like this: https://stackoverflow.com/a/2295681/2650104
Instead of splitting every character and doing a regex match operation

File: src/common/markdown.ts [hunk 2]
```diff
@@ -43,16 +43,49 @@ const defaultTextRender = markdown.renderer.rules.text;
 
 type MarkdownMention = Pick<User, 'id' | 'username'>;
 
+export const mentionSpecialCharacters = new RegExp('[^a-zA-Z0-9_@]', 'g');
+
+type ReplacedCharacters = string[];
+
+const getReplacedCharacters = (word: string): [string, ReplacedCharacters] => {
```
Comment (idoshamun) [inline src/common/markdown.ts:50]: I suggest adding here a short comment explaining about the logic of this function

**Review by idoshamun — APPROVED** [review]
I added a few ideas for improvement but you can merge at will :)
